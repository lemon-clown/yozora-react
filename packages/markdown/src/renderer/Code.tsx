import type { ICode } from '@yozora/ast'
import Code from '@yozora/react-code'
import React, { useContext } from 'react'
import { YozoraMarkdownContextType } from '../context/context'

export const YozoraCodeRenderer: React.FC<ICode> = props => {
  const { darken, preferLinenos, codeRunners } = useContext(YozoraMarkdownContextType)

  const { lang, meta } = props

  // Remove trailing line endings.
  const value: string = props.value.replace(/[\r\n]+$/, '')

  return (
    <Code
      lang={lang}
      value={value}
      meta={meta}
      runners={codeRunners}
      darken={darken}
      preferLinenos={preferLinenos}
    />
  )
}

YozoraCodeRenderer.displayName = 'YozoraCodeRenderer'
