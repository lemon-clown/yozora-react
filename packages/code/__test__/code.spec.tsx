import { render } from 'enzyme'
import React from 'react'
import Code from '../src'

describe('prop types', function () {
  beforeEach(() => {
    jest.spyOn(global.console, 'error').mockImplementation((...args) => {
      throw new Error(args.join(' '))
    })
  })

  it('render with custom className', () => {
    const code = 'let a = 1'
    const className = 'custom-code'
    const wrapper = render(<Code className={className} value={code} />)
    expect(wrapper.hasClass(className)).toBeTruthy()
  })

  it('value is required', () => {
    for (const value of [undefined, null] as any[]) {
      expect(() => {
        render(<Code value={value} />)
      }).toThrow(/The prop `value` is marked as required/i)
    }
  })
})

describe('snapshot', function () {
  const code = `
    function () {
      const a = 1;
      const b = 2;
      const c = a * a + b * b;
      return <span>Answer: {c}</span>
    }
  `.trim()
  const lang = 'jsx'

  it('live', () => {
    const wrapper = render(<Code meta="live maxlines=4" lang={lang} value={code} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('embed', () => {
    const wrapper = render(<Code meta="embed" lang={lang} value={code} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('literal', () => {
    const wrapper = render(
      <Code
        meta={`literal {1-2,2-1,4} title="/home/demo/a.tsx" collapsed`}
        lang={lang}
        value={code}
      />,
    )
    expect(wrapper).toMatchSnapshot()
  })
})
