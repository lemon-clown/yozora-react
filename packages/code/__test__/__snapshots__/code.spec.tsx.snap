// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot embed 1`] = `
<div
  class="yozora-code-embed"
/>
`;

exports[`snapshot literal 1`] = `
.c4 {
  display: inline-block;
  padding: 0.4rem;
  border: none;
  outline: none;
  background: transparent;
  cursor: pointer;
  color: hsl(0deg,0%,80%);
}

.c4:hover {
  color: hsl(0deg,0%,100%);
}

.c4:active {
  color: hsl(0deg,0%,90%);
}

.c1 {
  display: inline-block;
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  margin-left: 8px;
  background-color: #ed6c60;
}

.c2 {
  display: inline-block;
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  margin-left: 8px;
  background-color: #f7c151;
}

.c3 {
  display: inline-block;
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  margin-left: 8px;
  background-color: #64c856;
}

.c0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<div
  class="yozora-code-literal"
>
  <div
    class="yozora-code-literal__toolbar"
  >
    <span
      class="c0 yozora-common-light-buttons"
    >
      <span
        class="c1"
        title="close"
        type="close"
      />
      <span
        class="c2"
        title="minimize"
        type="minimize"
      />
      <span
        class="c3"
        title="maximize"
        type="maximize"
      />
    </span>
    <span
      class="yozora-code-literal__title"
      title="/home/demo/a.tsx"
    >
      /home/demo/a.tsxÂ  | 6 lines.
    </span>
    <span
      class="yozora-code-literal__copy-button"
    >
      <button
        aria-label="Copy to clipboard"
        class="c4 yozora-common-copy-button"
        color="hsl(0deg, 0%, 80%)"
        type="button"
      >
        copy
      </button>
    </span>
  </div>
  <code
    class="yozora-code-literal__content"
  >
    <pre>
      <div
        class="yozora-code-highlighter prism-code language-jsx"
        style="color:#d4d4d4;background-color:none;line-height:1.6rem;max-height:0"
      >
        <div
          class="yozora-code-highlighter__linenos"
          style="width:2.2em"
        >
          <div
            class="yozora-code-highlighter__line yozora-code-highlighter__line--highlight"
          >
            <span>
              1
            </span>
          </div>
          <div
            class="yozora-code-highlighter__line yozora-code-highlighter__line--highlight"
          >
            <span>
              2
            </span>
          </div>
          <div
            class="yozora-code-highlighter__line"
          >
            <span>
              3
            </span>
          </div>
          <div
            class="yozora-code-highlighter__line yozora-code-highlighter__line--highlight"
          >
            <span>
              4
            </span>
          </div>
          <div
            class="yozora-code-highlighter__line"
          >
            <span>
              5
            </span>
          </div>
          <div
            class="yozora-code-highlighter__line"
          >
            <span>
              6
            </span>
          </div>
        </div>
        <div
          class="yozora-code-highlighter__codes"
        >
          <div
            class="yozora-code-highlighter__codes-wrapper"
          >
            <div
              class="yozora-code-highlighter__line yozora-code-highlighter__line--highlight token-line"
              style="color:#d4d4d4"
            >
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                function
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                (
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                )
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                {
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="yozora-code-highlighter__line yozora-code-highlighter__line--highlight token-line"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                const
              </span>
              <span
                class="token plain"
              >
                 a 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                =
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token number"
                style="color:rgb(181, 206, 168)"
              >
                1
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                ;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="yozora-code-highlighter__line token-line"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                const
              </span>
              <span
                class="token plain"
              >
                 b 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                =
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token number"
                style="color:rgb(181, 206, 168)"
              >
                2
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                ;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="yozora-code-highlighter__line yozora-code-highlighter__line--highlight token-line"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                const
              </span>
              <span
                class="token plain"
              >
                 c 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                =
              </span>
              <span
                class="token plain"
              >
                 a 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                *
              </span>
              <span
                class="token plain"
              >
                 a 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                +
              </span>
              <span
                class="token plain"
              >
                 b 
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                *
              </span>
              <span
                class="token plain"
              >
                 b
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                ;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="yozora-code-highlighter__line token-line"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                      
              </span>
              <span
                class="token keyword"
                style="color:rgb(86, 156, 214)"
              >
                return
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &lt;
              </span>
              <span
                class="token tag"
                style="color:rgb(86, 156, 214)"
              >
                span
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &gt;
              </span>
              <span
                class="token plain"
              >
                Answer
              </span>
              <span
                class="token operator"
                style="color:rgb(212, 212, 212)"
              >
                :
              </span>
              <span
                class="token plain"
              >
                 
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                {
              </span>
              <span
                class="token plain"
              >
                c
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                }
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &lt;/
              </span>
              <span
                class="token tag"
                style="color:rgb(86, 156, 214)"
              >
                span
              </span>
              <span
                class="token tag punctuation"
                style="color:rgb(106, 153, 85)"
              >
                &gt;
              </span>
              <span
                class="token plain"
              />
            </div>
            <div
              class="yozora-code-highlighter__line token-line"
              style="color:#d4d4d4"
            >
              <span
                class="token plain"
              >
                    
              </span>
              <span
                class="token punctuation"
                style="color:rgb(106, 153, 85)"
              >
                }
              </span>
            </div>
          </div>
        </div>
      </div>
    </pre>
  </code>
</div>
`;

exports[`snapshot live 1`] = `
.c4 {
  display: inline-block;
  padding: 0.4rem;
  border: none;
  outline: none;
  background: transparent;
  cursor: pointer;
  color: hsl(0deg,0%,80%);
}

.c4:hover {
  color: hsl(0deg,0%,100%);
}

.c4:active {
  color: hsl(0deg,0%,90%);
}

.c1 {
  display: inline-block;
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  margin-left: 8px;
  background-color: #ed6c60;
}

.c2 {
  display: inline-block;
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  margin-left: 8px;
  background-color: #f7c151;
}

.c3 {
  display: inline-block;
  box-sizing: border-box;
  height: 10px;
  width: 10px;
  border-radius: 50%;
  margin-left: 8px;
  background-color: #64c856;
}

.c0 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

<div
  class="yozora-code-live"
>
  <div
    class="yozora-code-live__toolbar"
  >
    <span
      class="c0 yozora-common-light-buttons"
    >
      <span
        class="c1"
        title="close"
        type="close"
      />
      <span
        class="c2"
        title="minimize"
        type="minimize"
      />
      <span
        class="c3"
        title="maximize"
        type="maximize"
      />
    </span>
    <span
      class="yozora-code-live__title"
      title=""
    >
      Â 
    </span>
    <span
      class="yozora-code-live__copy-button"
    >
      <button
        aria-label="Copy to clipboard"
        class="c4 yozora-common-copy-button"
        color="hsl(0deg, 0%, 80%)"
        type="button"
      >
        copy
      </button>
    </span>
  </div>
  <div
    class="yozora-code-live__main yozora-code-live__main--vertical"
  >
    <div
      class="yozora-code-live__editor"
    >
      <div
        class="yozora-code-editor yozora-code-editor"
        padding="10"
        style="line-height:1.6rem;tab-size:2"
      >
        <div
          class="yozora-code-editor__textarea-wrapper"
        >
          <div
            class="yozora-code-editor__textarea-linenos"
          />
          <textarea
            autocapitalize="off"
            autocomplete="off"
            autocorrect="off"
            class="yozora-code-editor__textarea"
            spellcheck="false"
            style="tab-size:2;white-space:pre;padding:0.5rem 0.6em 0 6px"
          >
            function () {
      const a = 1;
      const b = 2;
      const c = a * a + b * b;
      return &lt;span&gt;Answer: {c}&lt;/span&gt;
    }
          </textarea>
        </div>
        <pre
          aria-hidden="true"
          class="yozora-code-editor__previewer"
          style="tab-size:2;white-space:pre"
        >
          <div
            class="yozora-code-highlighter prism-code language-jsx"
            style="color:#d4d4d4;background-color:none;line-height:1.6rem;max-height:8rem;min-height:100%"
          >
            <div
              class="yozora-code-highlighter__linenos"
              style="width:2.2em"
            >
              <div
                class="yozora-code-highlighter__line"
              >
                <span>
                  1
                </span>
              </div>
              <div
                class="yozora-code-highlighter__line"
              >
                <span>
                  2
                </span>
              </div>
              <div
                class="yozora-code-highlighter__line"
              >
                <span>
                  3
                </span>
              </div>
              <div
                class="yozora-code-highlighter__line"
              >
                <span>
                  4
                </span>
              </div>
              <div
                class="yozora-code-highlighter__line"
              >
                <span>
                  5
                </span>
              </div>
              <div
                class="yozora-code-highlighter__line"
              >
                <span>
                  6
                </span>
              </div>
            </div>
            <div
              class="yozora-code-highlighter__codes"
            >
              <div
                class="yozora-code-highlighter__codes-wrapper"
              >
                <div
                  class="yozora-code-highlighter__line token-line"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    function
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    (
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    )
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    {
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="yozora-code-highlighter__line token-line"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    const
                  </span>
                  <span
                    class="token plain"
                  >
                     a 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    =
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token number"
                    style="color:rgb(181, 206, 168)"
                  >
                    1
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    ;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="yozora-code-highlighter__line token-line"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    const
                  </span>
                  <span
                    class="token plain"
                  >
                     b 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    =
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token number"
                    style="color:rgb(181, 206, 168)"
                  >
                    2
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    ;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="yozora-code-highlighter__line token-line"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    const
                  </span>
                  <span
                    class="token plain"
                  >
                     c 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    =
                  </span>
                  <span
                    class="token plain"
                  >
                     a 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    *
                  </span>
                  <span
                    class="token plain"
                  >
                     a 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    +
                  </span>
                  <span
                    class="token plain"
                  >
                     b 
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    *
                  </span>
                  <span
                    class="token plain"
                  >
                     b
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    ;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="yozora-code-highlighter__line token-line"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                          
                  </span>
                  <span
                    class="token keyword"
                    style="color:rgb(86, 156, 214)"
                  >
                    return
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &lt;
                  </span>
                  <span
                    class="token tag"
                    style="color:rgb(86, 156, 214)"
                  >
                    span
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &gt;
                  </span>
                  <span
                    class="token plain"
                  >
                    Answer
                  </span>
                  <span
                    class="token operator"
                    style="color:rgb(212, 212, 212)"
                  >
                    :
                  </span>
                  <span
                    class="token plain"
                  >
                     
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    {
                  </span>
                  <span
                    class="token plain"
                  >
                    c
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    }
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &lt;/
                  </span>
                  <span
                    class="token tag"
                    style="color:rgb(86, 156, 214)"
                  >
                    span
                  </span>
                  <span
                    class="token tag punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    &gt;
                  </span>
                  <span
                    class="token plain"
                  />
                </div>
                <div
                  class="yozora-code-highlighter__line token-line"
                  style="color:#d4d4d4"
                >
                  <span
                    class="token plain"
                  >
                        
                  </span>
                  <span
                    class="token punctuation"
                    style="color:rgb(106, 153, 85)"
                  >
                    }
                  </span>
                </div>
              </div>
            </div>
          </div>
        </pre>
      </div>
    </div>
    <div
      class="yozora-code-live__previewer yozora-code-live__previewer--center"
    >
      <div
        class="yozora-code-embed"
      />
    </div>
  </div>
</div>
`;
